### PR TITLE
Ability to select and launch debug configuration

### LINKED ISSUE #16613: Ability to select and start a launch using keyboard
Ability to select and start a launch using keyboard

### PR DESCRIPTION
**Pull request by michelkaporin**
Adds 'debug' keyword similar to 'task' and 'git' to command palette.  

Fixes #16613.

### CODE REVIEW

File: src/vs/workbench/parts/debug/browser/debugActions.ts [hunk 1]
```diff
@@ -210,6 +211,28 @@ export class RunAction extends StartAction {
...
+	static ID = 'workbench.action.debug.launch';
+	static LABEL = nls.localize('launchDebugConfiguration', "Select and Start Debugging");
+
+	constructor(id: string, label: string,
+		@IDebugService debugService: IDebugService,
+		@IKeybindingService keybindingService: IKeybindingService,
+		@ICommandService commandService: ICommandService,
+		@IWorkspaceContextService contextService: IWorkspaceContextService,
+		@IFileService fileService: IFileService,
+		@IQuickOpenService quickOpenService: IQuickOpenService
```
Comment (michelkaporin) [inline src/vs/workbench/parts/debug/browser/debugActions.ts]: Put private for quickOpenService

File: src/vs/workbench/parts/debug/browser/debugQuickOpen.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+			autoFocusFirstEntry: !!input
+		};
+	}
+
+	public getEmptyLabel(searchString: string): string {
+		if (searchString.length > 0) {
+			return nls.localize('noConfigurationsMatching', "No debug configurations matching");
+		}
+
+		return nls.localize('noConfigurationsFound', "No debug configurations found");
```
Comment (michelkaporin) [inline src/vs/workbench/parts/debug/browser/debugQuickOpen.ts]: Update label to also contain '. Please create a 'launch.json' file.'

File: src/vs/workbench/parts/debug/browser/debugActions.ts [hunk 3]
```diff
@@ -210,6 +211,28 @@ export class RunAction extends StartAction {
 	}
 }
 
+export class LaunchAction extends RunAction {
```
Comment (michelkaporin) [inline src/vs/workbench/parts/debug/browser/debugActions.ts]: Extend from abstractdebugAction

File: src/vs/workbench/parts/debug/browser/debugActions.ts [hunk 4]
```diff
@@ -210,6 +211,28 @@ export class RunAction extends StartAction {
...
+		@IWorkspaceContextService contextService: IWorkspaceContextService,
+		@IFileService fileService: IFileService,
+		@IQuickOpenService quickOpenService: IQuickOpenService
+	) {
+		super(id, label, debugService, keybindingService, commandService, contextService, fileService);
+		this.quickOpenService = quickOpenService;
+	}
+
+	public run(): TPromise<any> {
+		return TPromise.as(this.quickOpenService.show('debug '));
```
Comment (michelkaporin) [inline src/vs/workbench/parts/debug/browser/debugActions.ts]: Do not wrap as promise

File: src/vs/workbench/parts/debug/browser/debugActions.ts (hunk 3 above)
Comment (michelkaporin) [inline src/vs/workbench/parts/debug/browser/debugActions.ts]: Rename to SelectAndStartAction, also change the id, nls.id

File: src/vs/workbench/parts/debug/electron-browser/debug.contribution.ts [hunk 5]
```diff
@@ -127,6 +128,17 @@ registry.registerWorkbenchAction(new SyncActionDescriptor(EnableAllBreakpointsAc
 registry.registerWorkbenchAction(new SyncActionDescriptor(DisableAllBreakpointsAction, DisableAllBreakpointsAction.ID, DisableAllBreakpointsAction.LABEL), 'Debug: Disable All Breakpoints', debugCategory);
 registry.registerWorkbenchAction(new SyncActionDescriptor(ClearReplAction, ClearReplAction.ID, ClearReplAction.LABEL), 'Debug: Clear Debug Console', debugCategory);
 registry.registerWorkbenchAction(new SyncActionDescriptor(FocusReplAction, FocusReplAction.ID, FocusReplAction.LABEL), 'Debug: Focus Debug Console', debugCategory);
+registry.registerWorkbenchAction(new SyncActionDescriptor(LaunchAction, LaunchAction.ID, LaunchAction.LABEL, null, CONTEXT_NOT_IN_DEBUG_MODE), 'Debug: Launch Configuration', debugCategory);
```
Comment (michelkaporin) [inline src/vs/workbench/parts/debug/electron-browser/debug.contribution.ts]: Remove null and context for now

File: src/vs/workbench/parts/debug/browser/debugActions.ts (hunk 4 above)
Comment (michelkaporin) [inline src/vs/workbench/parts/debug/browser/debugActions.ts]: Add an isEnabled() such that this action is only enabled while you are not debugging

File: src/vs/workbench/parts/debug/browser/debugQuickOpen.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+'use strict';
```
Comment (michelkaporin) [inline src/vs/workbench/parts/debug/browser/debugQuickOpen.ts]: We no longer 'use strict', this is simply ignored

File: src/vs/workbench/parts/debug/browser/debugQuickOpen.ts [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+import { TPromise } from 'vs/base/common/winjs.base';
+import Quickopen = require('vs/workbench/browser/quickopen');
+import QuickOpen = require('vs/base/parts/quickopen/common/quickOpen');
+import Model = require('vs/base/parts/quickopen/browser/quickOpenModel');
+import { IQuickOpenService } from 'vs/platform/quickOpen/common/quickOpen';
+import { IDebugService } from 'vs/workbench/parts/debug/common/debug';
+import * as errors from 'vs/base/common/errors';
+
+class DebugEntry extends Model.QuickOpenEntry {
+	private debugService: IDebugService;
```
Comment (michelkaporin) [inline src/vs/workbench/parts/debug/browser/debugQuickOpen.ts]: Get rid of this privates

File: src/vs/workbench/parts/debug/browser/debugQuickOpen.ts [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+	public getAriaLabel(): string {
+		return nls.localize('entryAriaLabel', "{0}, debug", this.getLabel());
+	}
+
+	public run(mode: QuickOpen.Mode, context: Model.IContext): boolean {
+		if (mode === QuickOpen.Mode.PREVIEW) {
+			return false;
+		}
+		// Run selected debug configuration
+		this.debugService.createProcess(this.configurationName).done(undefined, errors.onUnexpectedError);
```
Comment (michelkaporin) [inline src/vs/workbench/parts/debug/browser/debugQuickOpen.ts]: Do it the other way around

File: src/vs/workbench/parts/debug/browser/debugQuickOpen.ts [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+		this.debugService = debugService;
+	}
+
+	public getAriaLabel(): string {
+		return nls.localize('debugAriaLabel', "Type a name of a launch configuration to run.");
+	}
+
+	public getResults(input: string): TPromise<Model.QuickOpenModel> {
+		const configurationNames = this.debugService.getConfigurationManager().getConfigurationNames()
+			.sort((a, b) => a.localeCompare(b))
```
Comment (michelkaporin) [inline src/vs/workbench/parts/debug/browser/debugQuickOpen.ts]: Do not sort

**Comment by isidorn** [comment]
Looks good, merging in 🍻

**Pull request merged**